### DISCUSSION TITLE
请问能否讲解一下clang对语法树的生成逻辑

### DISCUSSION THREAD

**Discussion by Aurora-1412** [arcsysu/SYsU-lang#38] [Q&A]
觉得这有助于编程，例如什么情况下会有ImplicitCastExpr，单靠面向样例可能会不断重构代码（或者if套if套个几层这种）
p.s. 个人总结的ImpliciCastExpr：出现在左值转右值处（但是变量外加一个括号仍然视为左值）；出现在数组处（大概有n个[]就一共有(n+1)个ImplicitCastExpr）
再比如ReturnStmt：目前一共有return表达式、数组元素、变量、(变量)四种形式，每种形式我只能对着样例编程）没有一个比较全局性的代码逻辑（嗯加else if了）

或者有无相应的文档orz

**Comment by Yorkking** [comment]
你可以参考实验二的语法分析中，哪里需要引入 `ImplicitCastExpr`.

**Comment by wu-kan** [comment]
只需要检查是否存在 `"isImplicit"` 和 `"implicit"` 即可，没有的都是显式的。

**Reply by SundayDream00** [reply]
> 只需要检查是否存在 `"isImplicit"` 和 `"implicit"` 即可，没有的都是显式的。

请问这些是哪里的属性或者函数呢？在语法树里好像没看到相关，可以稍微详细说明一下吗？还是不太懂怎么利用这些来形成一个通用的全局代码逻辑

**Reply by wu-kan** [reply]
就是 `clang` 生成的语法树，你在进入节点前先检查 json 节点是否包含着两个属性，是否为 `true`。

**Reply by Aurora-1412** [reply]
> "implicit

其实……语法树里并没有这两个属性吧

**Reply by Yorkking** [reply]
> > "implicit
> 
> 其实……语法树里并没有这两个属性吧

有的，可能你暂时没遇到？

**Reply by Yorkking** [reply]
> > > "implicit
> > 
> > 
> > 其实……语法树里并没有这两个属性吧
> 
> 有的，可能你暂时没遇到？

就是当测试样例定义的函数和 Clang 内置函数重名时，Clang 会自己插入 Implicit 结点。
